cpp/src/arrow/compute/row/row_encoder_internal.h

```diff
@@ -269,6 +272,190 @@ struct ARROW_EXPORT NullKeyEncoder : KeyEncoder {
+          validity, data.array.offset, data.array.length,
+          [&](int64_t i) {
+            ARROW_UNUSED(i);
+            child_lengthes.clear();
+            Offset list_length = offsets[i + 1] - offsets[i];
+            if (list_length == 0) {
+              lengths[index] += kExtraByteForNull + sizeof(Offset);
+              ++index;
+              return;
+            }
+            child_lengthes.resize(list_length, 0);
+            tmp_child_data.SetSlice(offsets[i], list_length);
+            this->element_encoder_->AddLength(ExecValue{tmp_child_data}, list_length,
+                                              child_lengthes.data());
+            lengths[index] += kExtraByteForNull + sizeof(Offset);
+            for (int32_t j = 0; j < list_length; j++) {
+              lengths[index] += child_lengthes[j];
+            }
+            ++index;
+          },
+          [&]() {
+            lengths[index] = kExtraByteForNull + sizeof(Offset);
+            ++index;
+          });
+    } else {
+      const auto& list_scalar = data.scalar_as<BaseListScalar>();
+      int32_t accum_length = 0;
+      // Counting the size of the encoded list if the list is valid
+      if (list_scalar.is_valid && list_scalar.value->length() > 0) {
+        auto element_count = static_cast<int32_t>(list_scalar.value->length());
+        // Counting the size of the encoded list
+        std::vector<int32_t> child_lengthes(element_count, 0);
+        this->element_encoder_->AddLength(ExecValue{*list_scalar.value->data()},
+                                          element_count, child_lengthes.data());
+        for (int32_t i = 0; i < element_count; i++) {
+          accum_length += child_lengthes[i];
+        }
+      }
+      for (int64_t i = 0; i < batch_length; i++) {
+        lengths[i] += kExtraByteForNull + sizeof(Offset) + accum_length;
+      }
+    }
+  }
+
+  void AddLengthNull(int32_t* length) override {
+    *length += kExtraByteForNull + sizeof(Offset);
+  }
+
+  Status Encode(const ExecValue& data, int64_t batch_length,
+                uint8_t** encoded_bytes) override {
+    auto handle_null_value = [&encoded_bytes]() {
+      auto& encoded_ptr = *encoded_bytes++;
+      *encoded_ptr++ = kNullByte;
+      util::SafeStore(encoded_ptr, static_cast<Offset>(0));
+      encoded_ptr += sizeof(Offset);
+    };
+    auto handle_valid_value = [&encoded_bytes,
+                               this](const ArraySpan& child_array) -> Status {
+      auto& encoded_ptr = *encoded_bytes++;
+      *encoded_ptr++ = kValidByte;
+      util::SafeStore(encoded_ptr, static_cast<Offset>(child_array.length));
+      encoded_ptr += sizeof(Offset);
+      // handling the child data
+      for (int64_t i = 0; i < child_array.length; i++) {
+        ArraySpan tmp_child_data(child_array);
+        tmp_child_data.SetSlice(child_array.offset + i, 1);
+        RETURN_NOT_OK(
+            this->element_encoder_->Encode(ExecValue{tmp_child_data}, 1, &encoded_ptr));
+      }
```

Comment:
Why do it one element at a time? Why not instead:
```suggestion
      RETURN_NOT_OK(
          this->element_encoder_->Encode(ExecValue{child_array}, child_array.length, &encoded_ptr));
```